### PR TITLE
HDDS-12468. Check for space availability for all dns while container creation in pipeline

### PR DESCRIPTION
**Pull request by siddhantsangwan**
## What changes were proposed in this pull request?
When finding a container for a block, the SCM needs to allocate a new container if the existing containers don't match requirements. Currently, the SCM just creates a new container without checking if the Datanodes in that particular pipeline have enough disk space (5 GB) for the new container.

This pull request adds a check for both Ratis and EC flows. The check happens at a single place in `ContainerManagerImpl`. If the SCM is unable to allocate a new container then a different pipeline needs to be tried - code to handle this already exists.

The logic for the check itself already exists in `SCMCommonPlacementPolicy#hasEnoughSpace` and I reuse that here. This existing logic has interesting behaviour for one particular case - it returns false if the exact amount of space is available. For example if the required space is 5 GB, and one datanode has exactly 5 GB, the method will return false. It seems like a defensive way of handling this corner case of disks being close to full. 

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-12468

## How was this patch tested?

Added unit tests.

Ready for review, draft while waiting for CI - https://github.com/siddhantsangwan/ozone/actions/runs/15757908756/job/44418185632

### CODE REVIEW

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/WritableECContainerProvider.java [hunk 1]
```diff
@@ -203,8 +203,14 @@ private ContainerInfo allocateContainer(ReplicationConfig repConfig,
 
     Pipeline newPipeline = pipelineManager.createPipeline(repConfig,
         excludedNodes, Collections.emptyList());
+    // the returned ContainerInfo should not be null (due to not enough space in the Datanodes specifically) because
```
Comment (ChenSammi) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/WritableECContainerProvider.java:206]: Could newPipeline be null when DN is close to full?

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineManager.java [hunk 2]
```diff
@@ -216,4 +216,13 @@ void reinitialize(Table<PipelineID, Pipeline> pipelineStore)
...
   void releaseWriteLock();
+
+  /**
+   * Checks whether all Datanodes in the specified pipeline have greater than the specified space, containerSize.
+   * @param pipeline pipeline to check
+   * @param containerSize the required amount of space
+   * @return false if all the volumes on any Datanode in the pipeline have space less than equal to the specified
+   * containerSize, otherwise true
+   */
+  boolean pipelineHasEnoughSpaceForNewContainer(Pipeline pipeline, long containerSize);
```
Comment (ChenSammi) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineManager.java]: pipelineHasEnoughSpaceForNewContainer -> HasEnoughSpace

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerManagerImpl.java [hunk 3]
```diff
@@ -346,14 +352,24 @@ public ContainerInfo getMatchingContainer(final long size, final String owner,
       synchronized (pipeline.getId()) {
         containerIDs = getContainersForOwner(pipeline, owner);
         if (containerIDs.size() < getOpenContainerCountPerPipeline(pipeline)) {
-          allocateContainer(pipeline, owner);
-          containerIDs = getContainersForOwner(pipeline, owner);
+          if (pipelineManager.pipelineHasEnoughSpaceForNewContainer(pipeline, containerSize)) {
+            allocateContainer(pipeline, owner);
+            containerIDs = getContainersForOwner(pipeline, owner);
+          } else {
+            LOG.debug("Cannot allocate a new container because pipeline {} does not have the required space {}.",
```
Comment (ChenSammi) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerManagerImpl.java:359]: Please add the if (LOG.isDebugEnabled) check.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerManager.java [hunk 4]
```diff
@@ -191,7 +191,8 @@ default ContainerInfo getMatchingContainer(long size, String owner,
    * @param owner - the user which requires space in its owned container
    * @param pipeline - pipeline to which the container should belong.
    * @param excludedContainerIDS - containerIds to be excluded.
-   * @return ContainerInfo for the matching container.
+   * @return ContainerInfo for the matching container, or null if a container could not be found and could not be
+   * allocated
```
Comment (ChenSammi) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerManager.java:194]: Use  @Nullable, which is very straightforward.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerManagerImpl.java (hunk 3 above)
Comment (siddhantsangwan) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerManagerImpl.java:359]: AFAIK the `LOG.isDebugEnabled` check isn't needed when using slf4j with "{}" style logging. The log string is evaluated after the debug method itself calls `isDebugEnabled()`.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/WritableECContainerProvider.java (hunk 1 above)
Comment (siddhantsangwan) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/WritableECContainerProvider.java:206]: Good point. I checked, pipeline will not be null. There's a `checkPipeline(pipeline)` method that checks for null and throws an exception instead.

**Comment by siddhantsangwan** [comment]
@ChenSammi thanks for the review. Addressed your comments in the latest commit.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerManagerImpl.java [hunk 5]
```diff
@@ -346,14 +352,24 @@ public ContainerInfo getMatchingContainer(final long size, final String owner,
       synchronized (pipeline.getId()) {
         containerIDs = getContainersForOwner(pipeline, owner);
         if (containerIDs.size() < getOpenContainerCountPerPipeline(pipeline)) {
-          allocateContainer(pipeline, owner);
-          containerIDs = getContainersForOwner(pipeline, owner);
+          if (pipelineManager.hasEnoughSpace(pipeline, containerSize)) {
```
Comment (nandakumar131) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerManagerImpl.java]: Can we move this check to `allocateContainer` method?

Comment (siddhantsangwan) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerManagerImpl.java]: Moving the check to `allocateContainer` causes too many unrelated test failures because it is being used in unit tests which don't care about pipeline state. I could move it but we'll need to fix those other tests then.

Comment (siddhantsangwan) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerManagerImpl.java]: Can fix them by doing something like this, what do you think?
```
diff --git a/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerManagerImpl.java b/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerManagerImpl.java
index 65b06bfd42..2e30b72888 100644
--- a/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerManagerImpl.java
+++ b/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerManagerImpl.java
@@ -243,6 +243,10 @@ private ContainerInfo createContainer(Pipeline pipeline, String owner)
   private ContainerInfo allocateContainer(final Pipeline pipeline,
                                           final String owner)
       throws IOException {
+    if (!pipelineManager.hasEnoughSpace(pipeline, containerSize)) {
+      return null;
+    }
+
     final long uniqueId = sequenceIdGen.getNextId(CONTAINER_ID);
     Preconditions.checkState(uniqueId > 0,
         "Cannot allocate container, negative container id" +
diff --git a/hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/TestContainerManagerImpl.java b/hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/TestContainerManagerImpl.java
index baa09cd854..df74709f93 100644
--- a/hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/TestContainerManagerImpl.java
+++ b/hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/TestContainerManagerImpl.java
@@ -84,7 +84,7 @@ public class TestContainerManagerImpl {
   private SequenceIdGenerator sequenceIdGen;
   private NodeManager nodeManager;
   private ContainerReplicaPendingOps pendingOpsMock;
-  private PipelineManager pipelineManager;
+  private MockPipelineManager pipelineManager;
 
   @BeforeAll
   static void init() {
@@ -142,6 +142,7 @@ public void testGetMatchingContainerReturnsNullWhenNotEnoughSpaceInDatanodes() t
     Pipeline pipeline = pipelineManager.getPipelines().iterator().next();
     // MockPipelineManager#hasEnoughSpace always returns false
     // the pipeline has no existing containers, so a new container gets allocated in getMatchingContainer
+    pipelineManager.setHasEnoughSpaceBehavior(false);
     ContainerInfo container = containerManager
         .getMatchingContainer(sizeRequired, "test", pipeline, Collections.emptySet());
     assertNull(container);
@@ -159,9 +160,9 @@ public void testGetMatchingContainerReturnsContainerWhenEnoughSpaceInDatanodes()
     long sizeRequired = 256 * 1024 * 1024; // 256 MB
 
     // create a spy to mock hasEnoughSpace to always return true
-    PipelineManager spyPipelineManager = spy(pipelineManager);
-    doReturn(true).when(spyPipelineManager)
-        .hasEnoughSpace(any(Pipeline.class), anyLong());
+    PipelineManager spyPipelineManager = pipelineManager;
+//    doReturn(true).when(spyPipelineManager)
+//        .hasEnoughSpace(any(Pipeline.class), anyLong());
 
     // create a new ContainerManager using the spy
     File tempDir = new File(testDir, "tempDir");
diff --git a/hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/pipeline/MockPipelineManager.java b/hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/pipeline/MockPipelineManager.java
index 78f1865d2b..3274de3d8a 100644
--- a/hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/pipeline/MockPipelineManager.java
+++ b/hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/pipeline/MockPipelineManager.java
@@ -48,6 +48,7 @@
 public class MockPipelineManager implements PipelineManager {
 
   private final PipelineStateManager stateManager;
+  private boolean hasEnoughSpaceBehavior = true;
 
   public MockPipelineManager(DBStore dbStore, SCMHAManager scmhaManager,
                              NodeManager nodeManager) throws IOException {
@@ -352,6 +353,10 @@ public boolean isPipelineCreationFrozen() {
 
   @Override
   public boolean hasEnoughSpace(Pipeline pipeline, long containerSize) {
-    return false;
+    return hasEnoughSpaceBehavior;
+  }
+
+  public void setHasEnoughSpaceBehavior(boolean hasEnoughSpace) {
+    hasEnoughSpaceBehavior = hasEnoughSpace;
   }
 }
```

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerManagerImpl.java (hunk 3 above)
Comment (sodonnel) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerManagerImpl.java:359]: Yea if the log parameters are just simple references to be passed, the `isDebugEnabled()` is no needed as the logger does it internally.

Comment (ChenSammi) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerManagerImpl.java:359]: Good to know that slf4j with "{}" will delay the string construction, thanks!

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerManagerImpl.java (hunk 5 above)
Comment (siddhantsangwan) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerManagerImpl.java]: There's no handling for `null` in the `SCMClientProtocolServer` code flow that calls `allocateContainer`, so there can be a null pointer exception if `allocateContainer` returns null. I think this code path is being used for benchmarks. Probably better not to make changes that impact this path in this pull request.

Comment (nandakumar131) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerManagerImpl.java]: @siddhantsangwan, let's do this in a separate PR.

**Comment by nandakumar131** [comment]
@siddhantsangwan, There is a integration test failure, can you take a look at it?

**Comment by siddhantsangwan** [comment]
> @siddhantsangwan, There is a integration test failure, can you take a look at it?

It's unrelated and flaky. Passed in my fork - https://github.com/siddhantsangwan/ozone/actions/runs/15845888336

**Comment by siddhantsangwan** [comment]
I re-triggered the failing test.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerManagerImpl.java (hunk 5 above)
Comment (siddhantsangwan) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerManagerImpl.java]: Created a follow up jira - https://issues.apache.org/jira/browse/HDDS-13338

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerManagerImpl.java [hunk 6]
```diff
@@ -80,6 +81,8 @@ public class ContainerManagerImpl implements ContainerManager {
   @SuppressWarnings("java:S2245") // no need for secure random
   private final Random random = new Random();
 
+  private final long containerSize;
```
Comment (sumitagrawl) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerManagerImpl.java]: rename to maxContainerSize as from config

Comment (siddhantsangwan) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerManagerImpl.java]: done

**Comment by siddhantsangwan** [comment]
CI has passed, merging on the basis of Sumit's LGTM.
